docker.go

```diff
@@ -58,6 +60,29 @@ func (g *g) dockerConnect() (*docker, error) {
 	return &docker{client: cli, log: g.log}, nil
 }
 
+// isExistingLocalImage it doesn't work perfectly.
+// If the input image is "docker.io/library/mongo:4.4" and existing image tag is
+// "mongo:4.4", it will return true(expected).
+// If the input image is "docker.io/circleci/mongo:4.4" and existing image tag is
+// "mongo:4.4"(not "circleci/mongo:4.4"), it will still return true(unexpected).
+// We should handle this corner case.
+func (d *docker) isExistingLocalImage(ctx context.Context, image string) (bool, error) {
+	images, err := d.client.ImageList(ctx, types.ImageListOptions{All: true})
+	if err != nil {
+		return false, fmt.Errorf("can't list image: %w", err)
+	}
+
+	for _, img := range images {
+		for _, repoTag := range img.RepoTags {
+			if strings.HasSuffix(image, repoTag) {
+				return true, nil
+			}
```

Comment:
The problem you are describing in the comments relates to short image names, without docker organization. From my experience so far, if an image has a short name, like `mongo`, it is safe to convert it to `library/mongo`. If we add the following piece of code just before these 3 lines, the whole edge case of `mongo` vs `circleci/mongo` should go away:

```go
			if !strings.Contains(repoTag, "/") {
				repoTag = "library/" + repoTag
			}
```

Making this change will allow to reduce the complexity of the code by removing the need for an optional retry that you added to mitigate false-positives, regular expression matches and probably the entire `prepareContainer` function.

The good thing is that you added a test that covers this edge case, and we will know if my suggested change breaks anything or not.

Reply:
Thanks. It is the better way to resolve the edge case.